File: app/src/main/java/com/example/notes/recycler/NotesAdapter.java
package com.example.notes.recycler;

import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.notes.repo.NoteEntity;

import java.util.ArrayList;
import java.util.List;

public class NotesAdapter extends RecyclerView.Adapter<NoteViewHolder> {
    private List<NoteEntity> list = new ArrayList<>();
    private OnItemClickListener onItemClickListener;
    private OnItemDeleteListener onItemDeleteListener;

    public void setList(List<NoteEntity> list) {
        this.list = list;
        notifyDataSetChanged();
    }

    public interface OnItemClickListener {
        void onItemClick(NoteEntity note);
    }
    public interface OnItemDeleteListener {
        void onItemDelete(NoteEntity note);
    }

    public void setOnItemClickListener(OnItemClickListener onItemClickListener) {
        this.onItemClickListener = onItemClickListener;
    }
    public void setOnItemDeleteListener(OnItemDeleteListener onItemDeleteListener) {
        this.onItemDeleteListener = onItemDeleteListener;
    }

    @NonNull
    @Override
    public NoteViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new NoteViewHolder(parent, onItemClickListener, onItemDeleteListener);
    }

    @Override
    public void onBindViewHolder(@NonNull NoteViewHolder holder, int position) {
        holder.bind(list.get(position));
    }

    @Override
    public int getItemCount() {
        return list.size();
    }
}

File: app/src/main/java/com/example/notes/recycler/NotesAdapter.kt
package com.example.notes.recycler

import androidx.recyclerview.widget.RecyclerView
import com.example.notes.repo.NoteEntity
import android.view.ViewGroup

class NotesAdapter : RecyclerView.Adapter<NoteViewHolder>() {
    private var list: MutableList<NoteEntity> = mutableListOf()
    private var onItemListener: OnItemListener? = null

    fun setList(list: List<NoteEntity>) {
        this.list = list.toMutableList()
        notifyDataSetChanged()
    }

    fun addItem(note: NoteEntity) {
        list.add(note)
        notifyItemInserted(itemCount-1)
    }

    fun removeItem(note: NoteEntity) {
        val index = list.indexOf(note)
        list.remove(note)
        notifyItemRemoved(index)
    }

    interface OnItemListener {
        fun onItemClick(note: NoteEntity?)
        fun onItemDelete(note: NoteEntity?)
    }

    fun setOnItemListener(onItemListener: OnItemListener?) {
        this.onItemListener = onItemListener
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
        NoteViewHolder(parent, onItemListener)

    override fun onBindViewHolder(holder: NoteViewHolder, position: Int) {
        holder.bind(list[position])
    }

    override fun getItemCount() = list.size
}